File: DataStructures/src/main/java/com/zhangdd/stack/PolandNotation.java
package com.zhangdd.stack;


import java.util.*;
import java.util.stream.Collectors;

/**
 * @author zhangdd on 2021/10/10
 */
public class PolandNotation {


    public String infix2suffix(String infixExpression) {
        //将中缀表达式转化成集合方便遍历
        char[] chars = infixExpression.toCharArray();
        List<String> elements = new ArrayList<>();

        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            builder.append(c);
            //当前字符是数字，且下一位未越界且也是数字
            if (String.valueOf(c).matches("\\d+")
                    && i + 1 < chars.length && String.valueOf(chars[i + 1]).matches("\\d+")) {
            } else {
                elements.add(builder.toString());
                builder.delete(0, builder.length());
            }
        }


        //将中缀表达式转换为后缀表达式
        Stack<String> s1 = new Stack<>();
        List<String> s2 = new ArrayList<>();
        for (int i = 0; i < elements.size(); i++) {
            String s = elements.get(i);
            if (s.matches("\\d+")) {
                s2.add(s);
            } else if (s.equals("(")) {
                s1.push(s);
            } else if (s.equals(")")) {
                while (!s1.peek().equals("(")) {
                    s2.add(s1.pop());
                }
                s1.pop();
            } else {
                while (true) {
                    if (s1.isEmpty() || s1.peek().equals("(")) {
                        s1.push(s);
                        break;
                    } else if (OpeEnum.obtainOpe(s).getPriority() >
                            OpeEnum.obtainOpe(s1.peek()).getPriority()) {
                        s1.push(s);
                        break;
                    } else {
                        s2.add(s1.pop());
                    }
                }
            }
        }

        while (!s1.isEmpty()) {
            s2.add(s1.pop());
        }
        return String.join(" ", s2);
    }


    public int calculate(String suffixExpression) {
        String[] values = suffixExpression.split(" ");
        Stack<String> stack = new Stack<>();
        for (int i = 0; i < values.length; i++) {
            String value = values[i];
            if (isNum(value)) {
                stack.push(value);
            } else {
                int num2 = Integer.parseInt(stack.pop());
                int num1 = Integer.parseInt(stack.pop());
                int calculateValue = doCalculator(num1, num2, value);
                stack.push(String.valueOf(calculateValue));
            }
        }

        return Integer.parseInt(stack.peek());
    }

    @SuppressWarnings("DuplicatedCode")
    private int doCalculator(int num1, int num2, String ope) {
        OpeEnum opeEnum = OpeEnum.obtainOpe(ope);
        int result = 0;
        switch (opeEnum) {
            case ADD:
                result = num1 + num2;
                break;
            case SUB:
                result = num1 - num2;
                break;
            case MULTI:
                result = num1 * num2;
                break;
            case DIV:
                result = num1 / num2;
                break;
        }
        return result;
    }

    private boolean isNum(String s) {
        return s.matches("\\d+");
    }

    public enum OpeEnum {
        ADD("+", 1),
        SUB("-", 1),
        MULTI("*", 2),
        DIV("/", 2),
        ;

        private static final Map<String, OpeEnum> values = Arrays.stream(OpeEnum.values()).collect(Collectors.toMap(OpeEnum::getOpe, item -> item));

        private String ope;
        private int priority;

        OpeEnum(String ope, int priority) {
            this.ope = ope;
            this.priority = priority;
        }

        public String getOpe() {
            return ope;
        }

        public int getPriority() {
            return priority;
        }

        public Map<String, OpeEnum> getValues() {
            return values;
        }

        public static OpeEnum obtainOpe(String ope) {
            OpeEnum opeEnum = values.get(ope);
            if (opeEnum == null) {
                throw new RuntimeException("操作符不存在");
            }
            return opeEnum;
        }
    }

}

File: python/DataStructures/stack/PolandNotation.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/1/18
波兰表达式
"""
import enum


class OpeEnum(enum.Enum):
    def __init__(self, ope, priority):
        self.ope = ope
        self.priority = priority

    ADD = ('+', 1)
    SUB = ('-', 1)
    MULTI = ('*', 2)
    DIV = ('/', 2)

    @staticmethod
    def obtain_ope(ope):
        for item in OpeEnum:
            if item.ope == ope:
                return item
        raise Exception('不支持的操作符')

    @staticmethod
    def is_ope(ope):
        for item in OpeEnum:
            if item.ope == ope:
                return True
        return False


class PolandNotation:
    def __init__(self):
        pass

    def infix2suffix(self, infix_express):

        stack1 = []
        stack2 = []
        index = -1
        while True:
            index += 1
            if index >= len(infix_express):
                break

            if infix_express[index].isdigit():
                stack2.append(infix_express[index])
            elif infix_express[index] == '(':
                stack1.append(infix_express[index])
            elif infix_express[index] == ')':
                flag = True
                while flag and len(stack1) > 0:
                    tmp = stack1.pop()
                    if tmp == '(':
                        flag = False
                    else:
                        stack2.append(tmp)
            else:
                flag = True
                while flag:
                    if len(stack1) <= 0 or stack1[-1] == '(':
                        stack1.append(infix_express[index])
                        flag = False
                    elif OpeEnum.obtain_ope(infix_express[index]).priority > OpeEnum.obtain_ope(stack1[-1]).priority:
                        stack1.append(infix_express[index])
                        flag = False
                    else:
                        tmp = stack1.pop()
                        stack2.append(tmp)

        while True:
            if len(stack1) <= 0:
                break
            tmp = stack1.pop()
            stack2.append(tmp)

        return stack2

    def calculate(self, suffix_expression):
        """
        后缀表达式计算
        :param suffix_expression:
        :return:
        """

        num_stack = []
        index = -1
        while True:
            index += 1
            if index >= len(suffix_expression):
                break

            if OpeEnum.is_ope(suffix_expression[index]):
                num2 = num_stack.pop()
                num1 = num_stack.pop()
                result = self.do_calculator(num1, num2, suffix_expression[index])
                num_stack.append(result)
            else:
                num_stack.append(suffix_expression[index])

        return num_stack.pop()

    def do_calculator(self, num1, num2, ope):
        ope_enum = OpeEnum.obtain_ope(ope)
        if ope_enum is OpeEnum.ADD:
            return int(num1) + int(num2)
        elif ope_enum is OpeEnum.SUB:
            return int(num1) - int(num2)
        elif ope_enum is OpeEnum.MULTI:
            return int(num1) * int(num2)
        else:
            return int(num1) / int(num2)


poland_notation = PolandNotation()

result = poland_notation.infix2suffix('1+((2+3)*4)-5')
print(result)

result = poland_notation.calculate(result)
print(result)
